### ISSUE TITLE
About helper.py

### ISSUE THREAD

**Issue by fzh0917** [amoudgl/pygoturn#21]
Hello, @amoudgl , I am reading your code one line by one line recently. I am having some troubles reading the `helper.py` file. I have a bit of confusion about those functions and variables' meaning. Through the paper and supplementary material, I think that I have understood completely the principle of `GOTURN`, but the reason why I am feeling confused is that I think the logic of image preprocessing is not so complicated. 
So, can you make some explanations for me? thank you very much!

**Comment by fzh0917** [comment]
I am a non-English speaking person, so if you can't understand my meaning completely, please tell me immediately, I will try my best to express what I want to say. Thank you again!

**Comment by fzh0917** [comment]
I have made some tries, such as displaying bounding boxes and images variables by using OpenCV, printing coordinates, widths, heights and shapes...etc. However, After I do that, I have more confusions about those variables' meaning, which are not what I have understood. So, I need your help very much. Thank you!

**Comment by fzh0917** [comment]
Actually, The functions I can't understand are `shift_crop_training_sample(sample, bb_params)`, `crop_sample(sample)`, `cropPadImage(bbox_tight, image)` and `computeCropPadImageLocation(bbox_tight, image)`. 
I know a part of the logic of the function `computeCropPadImageLocation(bbox_tight, image)`, but after reading those variables, such as `left_half`, `right_half` and `top_half`...etc, I have a question which is that it's necessary to write the logic so complicated?

**Comment by amoudgl** [comment]
Hi @fzh0917, I adapted most of these methods from PY-GOTURN repo. In my `shift_crop_training_sample` method, I randomly shift previous box location with smooth motion model as described in the paper and then return the shifted box location _with respect to_ cropped image. 

In order to preserve the location of shifted box in original image dimensions, I have params like `edge_spacing_x`, `edge_spacing_y` which gives me distance from x and y axis in original image etc. Since I directly used `boundingbox.py` from [PY-GOTURN](https://github.com/nrupatunga/PY-GOTURN/blob/goturn-dev/goturn/helper/BoundingBox.py) repo, I didn't change its nomenclature. `search_region` gives me the region (box + context padded image) from original image which was cropped without any resizing. This is useful since we do not maintain aspect ratio in GOTURN while resizing, hence for unscaling the network output, we need the original search region dimensions.

All these optional parameters [`opts` in `helper.py`] are _only_ used while inference for doing in-place conversion of bounding box in the original image dimensions, see [here](https://github.com/amoudgl/pygoturn/blob/master/src/goturn.py#L109). Network just returns the location of box with respect to resized crop. This exact same framework was followed in the original GOTURN and PY-GOTURN repositories.

Image cropping methods in `helper.py`, namely `cropPadImage` and `computeCropPadImageLocation` were directly copied from PY-GOTURN (see [here](https://github.com/nrupatunga/PY-GOTURN/blob/goturn-dev/goturn/helper/image_proc.py)). I followed this in order to exactly reproduce the GOTURN crop methodology, PY-GOTURN is a direct replication of GOTURN C++ code to Python. I tested them and they were working fine, so didn't bother to change them. :)

**Comment by fzh0917** [comment]
Hi, @amoudgl , thank you very much for your detailed explanations in the last week. I have understood all my questions now. However, there is a new question that comes to my mind today, which is that why we need change the orders of images' channels from `B, G, R` to `R, G, B` in preprocessing images? What is the purpose to do this? Thanks.

**Comment by amoudgl** [comment]
I am using a pytorch pretrained model which expects inputs to be RGB images with a specific mean and std: https://pytorch.org/docs/stable/torchvision/models.html 

Thus, passing BGR images would not be appropriate and it leads to suboptimal results.

**Comment by fzh0917** [comment]
OK.I get it. However, I used some videos taken from the `OTB` dataset to test the model, and I found that the performances were better when not calling `bgr2rgb` function. 
Well, maybe they are only some special samples.
Thanks very much for your patient help.
